week03_model_free/homework.ipynb

```diff
@@ -766,6 +766,49 @@
    },
    "outputs": [],
    "source": []
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "collapsed": true
+   },
+   "outputs": [],
+   "source": [
+    "%%writefile ../run_display_servers.sh\n",
+    "# OPEN PORT 6080!\n",
+    "apt-get update\n",
+    "apt-get install -y python-tk\n",
+    "apt-get install -y x11vnc\n",
+    "x11vnc -R stop\n",
+    "x11vnc -bg -forever -display :1 -rfbport 5900\n",
+    "ls noVNC > /dev/null && rm -r noVNC\n",
+    "git clone https://github.com/novnc/noVNC.git\n",
+    "cp noVNC/vnc.html noVNC/index.html\n",
+    "# Do not work if it run in Jupyter\n",
+    "#head -n 166 noVNC/utils/launch.sh > noVNC/utils/bglaunch.sh\n",
+    "#chmod ago+x noVNC/utils/bglaunch.sh\n",
+    "#noVNC/utils/bglaunch.sh --vnc localhost:5900 --listen 6080\n",
+    "# Instead run in a terminal window ...\n",
+    "# noVNC/utils/launch.sh --vnc localhost:5900 --listen 6080\n",
+    "# ... and go to http://servername:6080 page"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "is_not_installed = False # Delete this line if you want to see the display\n",
+    "try:\n",
+    "    is_not_installed == True\n",
+    "except NameError:\n",
+    "    is_not_installed = True\n",
```

Comment:
This is bad. Instructions are unclear, double negation in `is_not_installed = False`, branches depending on `try...except NameError`.

Reply:
This instructuon is aimed
1) to prevent double running of x11vnc
2) to prevent unwanted x11vnc running by "Run all", "Run below", a streak of "Shift+Enter" etc. if a user does not even know about this feature

Thus the "is_not_installed = False \# Delete this line if you want to see the display " achive the 2) aim.
After deletion the "is_not_installed = False ..." (and kernel restarting or "del is_not_installed" if line was run), enviroment will not know about "is_not_installed" and will set it to "True" it only once in one kernel session, the installing will set it to "False". Thus the 1) aim will be achived.

Double negation is all about avoiding not(...) function in line  "if is_not_installed".

Reply:
Behavior of this code is complicated and depends on a lot of state. In order to understand it, I have to think through each possible case — basically trace each execution path in my head. This is not OK.

How about this:

```python
enable_visualization = True  # set to False if you don't want visualization

try:
    x11vnc_running
except NameError:
    x11vnc_running = False

if enable_visualization and not x11vnc_running:
    ...
```